js/components/contact_download.jsx

```diff
@@ -0,0 +1,165 @@
+import React, { Component } from 'react';
+import { Container } from 'flux/utils';
+import { requestActions } from 'actions';
+import ExcellentExport from 'excellentexport';
+import agencyComponentStore from '../stores/agency_component';
+
+/**
+ * A download button for getting all agency component contact information as CSV.
+ */
+class ContactDownloadButton extends Component {
+  static getStores() {
+    return [agencyComponentStore];
+  }
+
+  static calculateState() {
+    const { agencyComponents } = agencyComponentStore.getState();
+    return {
+      agencyComponents,
+    };
+  }
+
+  /**
+   * Get the agency_component fields that contain contact information.
+   */
+  static getFoiaPersonnelFields() {
+    return ['foia_officers', 'public_liaisons', 'service_centers'];
+  }
+
+  componentDidMount() {
+    // If there is any agency data, assume all the data is fetched.
+    if (this.state.agencyComponents.size) {
+      return;
+    }
+
+    // Pre-fetch the list of agency components, re-purposing the "finder" code.
+    const includeReferenceFields = {
+      agency_component: ['title', 'agency', 'submission_address', 'submission_fax', 'website'],
+      agency: ['name'],
+    };
+    ContactDownloadButton.getFoiaPersonnelFields().forEach((personnelField) => {
+      includeReferenceFields.agency_component.push(personnelField);
+      includeReferenceFields[personnelField] = ['name', 'title', 'email', 'phone'];
+    });
+    requestActions.fetchAgencyFinderData(includeReferenceFields);
+  }
+
+  /**
+   * Download the CSV file.
+   */
+  performDownload() {
+    return ExcellentExport.convert({
+      anchor: this.downloadButton,
+      filename: 'FOIA-Agency-Contacts',
+      format: 'csv',
+    }, [{
+      name: 'FOIA Agency Contacts',
+      from: {
+        array: this.rows,
+      },
+    }]);
+  }
+
+  /**
+   * Event handler for when the download button is triggered.
+   */
+  handleClick() {
+    return (this.rows) ? this.performDownload() : false;
+  }
+
+  /**
+   * Called each time a batch of agency components finished loading.
+   */
+  render() {
+    const progress = agencyComponentStore.getState().agencyFinderDataProgress;
+
+    // If this is not the last batch, just update the button text.
+    if (progress < 100) {
+      if (this.downloadButton) {
+        this.downloadButton.innerHTML = `Please wait - ${progress}%`;
```

Comment:
Instead of manipulating the DOM, this change should be reflected in the html you are returning e.g.

```
const downloadButtonLabel  = `Please wait - ${progress}%`;
// ...

return <a>{ downloadButtonLabel }</a>;
```